fictitous-commit

GuyPardo opened the pull request.

Commit:
fictitous-commit

Commit:
more friendly printing

during QiskitExperimentDensityMat : sweep and labber_read

Commit:
add LAbber_utils, nicer printing

Commit:
nicer printing

ori-pardo commented on Labber_util.py:
```diff
@@ -0,0 +1,37 @@
+"""
+Created on Mon Jun 20 15:00:15 2022
+
+@author: owner
```
Who is owner?

GuyPardo replied:
probably me.. fixed.

GuyPardo replied:
I think it was generated automatically when I still worked with Spyder

ori-pardo commented on Labber_util.py:
```diff
@@ -0,0 +1,37 @@
+# -*- coding: utf-8 -*-
```
This CR should be on all the files in the projects, but it seems that some of them are missing from the `Files changed` category so I can't review them in this merge request.
Regardless, many comments in this review would probably apply to all files in the project.

ori-pardo replied:
File names should be in `snake_case` or `CamelCase` (only for classes), but not both.

ori-pardo commented on Labber_util.py:
```diff
@@ -0,0 +1,37 @@
+"""
+
+import Labber as lb
+import os
```
Two empty lines before function declaration outside of a class. Fix everywhere
(or hit Ctl+Alt+L in PyCharm)

ori-pardo commented on Labber_util.py:
```diff
@@ -0,0 +1,37 @@
+
+import Labber as lb
+import os
+def log_exists(log_name):
```
Missing docstring. Fix everywhere.
example:
```python
def do_stuff(a, b, c):
    """
    Do stuff.
    @param a: ...
    @param b: ...
    @param c: ...
    @return: C{str}
    """
```
No need for `@return: ` if the function does not return anything

ori-pardo commented on Labber_util.py:
```diff
@@ -0,0 +1,37 @@
+    lLog = [dict(name='name', vector = False)]
+    templog = lb.createLogFile_ForData('temp', lLog)
+    path = templog.getFilePath(None)
+    folder = os.path.dirname(path)
+    log_name = f'{log_name}.hdf5'
+    new_file_path = os.path.join(folder, log_name)
```
```suggestion
    new_file_path = os.path.join(os.path.dirname(path), f'{log_name}.hdf5')
```

GuyPardo replied:
I think this is less readable

ori-pardo commented on Labber_util.py:
```diff
@@ -0,0 +1,37 @@
+    #create a temporary log:
+    lLog = [dict(name='name', vector = False)]
+    templog = lb.createLogFile_ForData('temp', lLog)
+    path = templog.getFilePath(None)
```
`path` is to broad a name. What is it a path for?

GuyPardo replied:
fixed

ori-pardo commented on Labber_util.py:
```diff
@@ -0,0 +1,37 @@
+
+
+def get_log_name(log_name):
+    i=1
```
There is probably a better name for `i` here.

GuyPardo replied:
fixed

ori-pardo commented on Labber_util.py:
```diff
@@ -0,0 +1,37 @@
+    log_name_temp = log_name
+    while log_exists(log_name_temp):
+        log_name_temp = f'{log_name}__{i}'
+        i=i+1
```
```suggestion
        i += 1
```
Fix everywhere.

GuyPardo replied:
this is an aesthetic choice and I think i=i+1 is more readable

ori-pardo commented on Labber_util.py:
```diff
@@ -0,0 +1,37 @@
+
+def get_log_name(log_name):
+    i=1
+    log_name_temp = log_name
+    while log_exists(log_name_temp):
+        log_name_temp = f'{log_name}__{i}'
```
`log_name_temp` is redundant.
```suggestion
    while log_exists(f'{log_name}__{i}'):
```
and return `f'{log_name}__{i}` at the end.
Also, `'{log_name}__{i}'` should probably be a constant
```python
# declaration
LOG_NAME_FORMAT = '{log_name}__{counter}`
# ...

# usage
LOG_NAME_FORMAT.format(log_name="blabla", counter=3)
```

GuyPardo replied:
this does not produce the same behavior as the original code:
I want that to get just log_name (without any integer suffix) in the case that does not already exist in the database
I did fix  constant format thing.

ori-pardo commented on Labber_util.py:
```diff
@@ -0,0 +1,37 @@
+    return log_name_temp    
+
+def open_log(log_name):
+    lLog = [dict(name='name', vector = False)]
+    templog = lb.createLogFile_ForData('temp', lLog)
+    path = templog.getFilePath(None)
+    folder = os.path.dirname(path)
+    log_name = f'{log_name}.hdf5'
+    new_file_path = os.path.join(folder, log_name)
```
This code is duplicated in `log_exists`. Make it a function.

GuyPardo replied:
fixed

ori-pardo commented on experiment_manager.py:
```diff
@@ -8,9 +8,12 @@
 from general_utils import enumerated_product
 
-sys.path.append(os.path.abspath(r"G:\My Drive\guy PHD folder\util"))
+#sys.path.append(os.path.abspath(r"G:\My Drive\guy PHD folder\util"))
```
Code does not belong in a comment. Fix Everywhere.

ori-pardo commented on experiment_manager.py:
```diff
@@ -8,9 +8,12 @@
 from dataclasses import dataclass
 import Labber
 import itertools as iter
+
+from tictoc import tic, toc
+
 from general_utils import enumerated_product
 
-sys.path.append(os.path.abspath(r"G:\My Drive\guy PHD folder\util"))
+#sys.path.append(os.path.abspath(r"G:\My Drive\guy PHD folder\util"))
 import Labber_util as lu
```
Reformat imports.
Use `Ctl`+`Alt`+`L` to reformat the file and `Ctl`+`Alt`+`M` to optimize imports.

GuyPardo replied:
fixed (but ctl+alt+M= doesn't work in my version)

ori-pardo commented on experiment_manager.py:
```diff
@@ -283,6 +286,10 @@ def sweep(self, config, save_to_labber=True, labber_log_name=None):
             # save in python: #TODO
 
 
+
+
+
+
```
Too many spaces

GuyPardo replied:
fixed

ori-pardo commented on experiment_manager.py:
```diff
@@ -323,7 +330,11 @@ def wait_result(self, job):
         return result
 
     def one_dimensional_job(self, config: Config):
+        # input verification
```
Comments like this should be written in Imperative tense.

GuyPardo replied:
fixed

ori-pardo commented on experiment_manager.py:
```diff
@@ -351,18 +365,26 @@ def sweep(self, config):
             # do 1D sweep on the tracing parameter:
-            print(curr_config.param_list) #TODO only print sweep parameters
+            print('current configuration:')
+            for par in outer_variables.param_list:
```
What is `par`?

GuyPardo replied:
changed to param (it shouldn't conflict with param from the earlier loop right?)

ori-pardo commented on experiment_manager.py:
```diff
@@ -351,18 +365,26 @@ def sweep(self, config):
             # do 1D sweep on the tracing parameter:
-            print(curr_config.param_list) #TODO only print sweep parameters
+            print('current configuration:')
+            for par in outer_variables.param_list:
+                print(getattr(curr_config,par.name))
```
This should probably be a method of `Config`

GuyPardo replied:
but this is more specific than that since outer_variables and curr_config are not the same
here I want to print other current scalar values of the looped parameters

but anyway a method Config.get_names_list() is a good idea and I added it

ori-pardo commented on experiment_manager.py:
```diff
@@ -323,7 +330,11 @@ def wait_result(self, job):
+            raise ValueError("config must have exactly one iterable Parameter")
         variable_param = config.get_iterables()[0]
+        tic()
         circs = []
```
What is `circs`?

GuyPardo replied:
a list of qiskit.QuantumCircuit objects that each one of them is generated by child method get_circ(config)

ori-pardo commented on experiment_manager.py:
```diff
@@ -323,7 +330,11 @@ def wait_result(self, job):
+        # input verification
+        if not len(config.get_iterables()) == 1:
+            raise ValueError("config must have exactly one iterable Parameter")
         variable_param = config.get_iterables()[0]
```
Add a comment to explain what is special about the first element in the list.

GuyPardo replied:
fixed

ori-pardo commented on experiment_manager.py:
```diff
@@ -323,7 +330,11 @@ def wait_result(self, job):
 
     def one_dimensional_job(self, config: Config):
+        # input verification
+        if not len(config.get_iterables()) == 1:
```
```suggestion
        if len(config.get_iterables()) != 1:
```

GuyPardo replied:
fixed

ori-pardo commented on experiment_manager.py:
```diff
@@ -334,6 +345,9 @@ def one_dimensional_job(self, config: Config):
 
         job = config.backend.value.run(circs)
         self._async_results.append(job)
+        print('1D job sent in:')
+        toc()
+        print('')
```
I am confused.

GuyPardo replied:
this is what it does:
![image](https://user-images.githubusercontent.com/67901011/183855388-8300c856-b30e-476c-93a5-3ff9c45a535b.png)

ori-pardo commented on experiment_manager.py:
```diff
@@ -351,18 +365,26 @@ def sweep(self, config):
         # N-dimensional loop with itertools.product: # (actually N-1 )
 
+        counter = 1
+        num_jobs = 1
+        for var in outer_variables.param_list:
+            num_jobs=num_jobs*len(var.value)
```
This look like a method of the `Config` class

GuyPardo replied:
fixed

ori-pardo commented on experiment_manager.py:
```diff
@@ -351,18 +365,26 @@ def sweep(self, config):
+        for var in outer_variables.param_list:
+            num_jobs=num_jobs*len(var.value)
 
         for indices, vals in enumerated_product(*outer_variables.get_values()):
```
I think `enumerated_product(*outer_variables.get_values())` is used in enough places to be a function. maybe a method of `Config` but i am not sure

GuyPardo replied:
I agree but this requires more thought because outer_variables is also used inside the loop.
I leave it for now but I add a TODO and it will be dealt with at some point..

ori-pardo commented on experiment_manager.py:
```diff
@@ -351,18 +365,26 @@ def sweep(self, config):
             self.sweep_configs.append(deepcopy(curr_config))
+            counter = counter + 1
 
     def get_observables(self,config:Config, density_matrix):
         # to be implemented in child class
```
There is a convention for that.
Either:
```suggestion
    def get_observables(self,config:Config, density_matrix):
        raise NotImplementedError()
```
or (In python 3)
```suggestion
    def get_observables(self,config:Config, density_matrix):
        ...
```

GuyPardo replied:
fixed

ori-pardo commented on experiment_manager.py:
```diff
@@ -432,8 +454,12 @@ def labber_read(self,config, labber_log_name=None):
+            toc()
+            print('')
             labber_trace = result["labber_trace"]
             #
             # print("labber trace")
```
Delete these

GuyPardo replied:
fixed

ori-pardo commented on experiment_manager.py:
```diff
@@ -432,8 +454,12 @@ def labber_read(self,config, labber_log_name=None):
+            print('1D observables read in:')
+            toc()
+            print('')
             labber_trace = result["labber_trace"]
```
No need for that variable.
Just return `result['labber_trace']`

GuyPardo replied:
fixed

Commit:
implemenet Config.is_constant

Commit:
add Config.get_parameters_dict() and Config.get_values_dict()

Commit:
fix Ori's comments on Labber_utils

Commit:
change Labber_util to labber_util

Commit:
fix Ori's comments on experiment_manager

except addinf docstrings evrywhere...
additionally fix typo in labber_utils

Commit:
change get_log_name

such that we only have to get the current Laber folder and create temp log once

Commit:
add docstrings

Commit:
added tags to labber read

Commit:
add shots, tags,

Commit:
add qiskit utils

Commit:
split_qiskit_to_new_file

ori-pardo commented on experiment_manager.py:
```diff
@@ -27,18 +30,24 @@ class Parameter:  # TODO - I realized this class can be used for output data as
 
-        if is_iterated == None:
+        if is_iterated is None:
             if isinstance(self.value, Iterable):
```
`self.is_iterated = isinstance(self.value, Iterable)` looks better in my opinion

ori-pardo commented on experiment_manager.py:
```diff
@@ -82,37 +144,78 @@ def set_parameter(self, **kwargs):
     def get_values(self):
+        """
+        TODO rename to get_values_list()?
+        get a list of the Config's Parameters values
```
```suggestion
        Get a list of the Config's Parameters values.
```
Fix everywhere.

ori-pardo commented on experiment_manager.py:
```diff
@@ -151,16 +272,40 @@ def get_labber_log_list(self):
+            iter_count = iter_count*len(var.value)
+        return iter_count
+
+#--------------- end of config class definition ----------------------#
```
I dont really like this type of comment, but it doens't hurt to have it.

ori-pardo commented on experiment_manager.py:
```diff
@@ -151,16 +272,40 @@ def get_labber_log_list(self):
+    # treat vector values that are not iterated
     for param in output_config_list[0].param_list:
         if not param.is_iterated:
             labber_dict[param.name] = np.array(labber_dict[param.name])
```
```suggestion
            labber_dict[param.name] = np.array(labber_dict[param.name])
            
```

ori-pardo commented on experiment_manager.py:
```diff
@@ -169,8 +314,9 @@ def get_labber_trace(output_config_list):
+    times with different configurations. and save the results to Labber
     """
 
     def __init__(self):
```
No need to implement a `__init__()` that does nothing

ori-pardo commented on experiment_manager.py:
```diff
@@ -284,179 +427,57 @@ def sweep(self, config, save_to_labber=True, labber_log_name=None):
-    def get_circ(self, config:Config):
-        #to be implemented in child class
-        raise  NotImplemented('get_circ method not implemented')
+class QiskitExperiment(AsyncExperiment):
```
You have a whole lot of classes in this module. I suggest splitting them to different files.
depending on how you want to do this, `QiskitExperiment` might not belong in this project but in another project that have this project as a dependency

ori-pardo commented on experiment_manager.py:
```diff
@@ -284,179 +427,57 @@ def sweep(self, config, save_to_labber=True, labber_log_name=None):
-        #to be implemented in child class
-        raise  NotImplemented('get_circ method not implemented')
+class QiskitExperiment(AsyncExperiment):
+    def get_circ(self, config: Config):
```
```suggestion
    def get_circuit(self, config: Config):
```

ori-pardo commented on labber_util.py:
```diff
@@ -0,0 +1,82 @@
+import Labber as lb
+import os
+
+def get_current_Labber_path():
```
```suggestion
def get_current_labber_path():
```

ori-pardo commented on labber_util.py:
```diff
@@ -0,0 +1,82 @@
+    :return:str the same name with additional numbers at the end as necessary to avoid overwrite.
+    """
+    labber_path = get_current_Labber_path()
+    LOG_NAME_FORMAT = '{log_name}__{integer_suffix}'
```
Constants belong at the beginning of the file.

ori-pardo commented on labber_util.py:
```diff
@@ -0,0 +1,82 @@
+    return log_name_temp
+
+
+def open_log(log_name):
```
This function is redundant IMO.

ori-pardo commented on qiskit_manager.py:
```diff
@@ -0,0 +1,234 @@
+import importlib
+import qiskit_utils as qu
+import experiment_manager
+importlib.reload(qu)
+importlib.reload(experiment_manager)
+from experiment_manager import *
```
If you have to do this, you are doing something wrong.

ori-pardo commented on qiskit_manager.py:
```diff
@@ -0,0 +1,234 @@
+
+
+
+class QiskitExperimentDensityMat(AsyncExperiment):
```
Shouldn't this class inherit from `QiskitExperiment`?

ori-pardo commented on qiskit_manager.py:
```diff
@@ -0,0 +1,234 @@
+
+    def run(self, config: Config):
+        """
+        runs the ciscuit :param config:Config with constant (not iterated) Parameters that are used in self.get_circ(
```
```suggestion
        :param config:Config with constant (not iterated) Parameters that are used in self.get_circ(
```

ori-pardo commented on qiskit_manager.py:
```diff
@@ -0,0 +1,234 @@
+
+    def wait_result(self, job):
+        """
+
```
Redundant empty space. Fix everywhere.

ori-pardo commented on qiskit_manager.py:
```diff
@@ -0,0 +1,234 @@
+            shots = config.shots.value
+        else:
+            shots = 1024
+
+
```
```suggestion

```
Fix everywhere.

ori-pardo commented on qiskit_manager.py:
```diff
@@ -0,0 +1,234 @@
+    def get_observables(self, config: Config, density_matrix):
+        raise NotImplementedError()
+        # should return an output Config object
+        pass
```
```suggestion
```

ori-pardo commented on qiskit_manager.py:
```diff
@@ -0,0 +1,234 @@
+        # returns a dict with output config list, and labber trace
+
+        # input verification
+        if not len(config.get_iterables()) == 1:
```
```suggestion
        if len(config.get_iterables()) != 1:
```

ori-pardo commented on qiskit_manager.py:
```diff
@@ -0,0 +1,234 @@
+            print('')
+            logfile.addEntry(result["labber_trace"])
+
+####################################################################
```
```suggestion
```

ori-pardo commented on qiskit_utils.py:
```diff
@@ -0,0 +1,173 @@
+
+    def __init__(self, T1, T2):
+        self.T1 = T1
+        if T2 > 2 * T1:
+            self.T2 = 2 * T1
+        else:
+            self.T2 = T2
```
```suggestion
        self.T2 = min(T2, 2 * T1)
```

ori-pardo commented on qiskit_utils.py:
```diff
@@ -0,0 +1,173 @@
+            self.T2 = T2
+
+
+def get_thermal_error_object(T1s, T2s, time_u1=0, time_u2=50, time_u3=100, time_cx=300, time_reset=1000,
```
Missing docstring. Fix everywhere.

ori-pardo commented on qiskit_utils.py:
```diff
@@ -0,0 +1,173 @@
+from qiskit.tools.visualization import plot_histogram, plot_state_city
+import qiskit.quantum_info as qi
+
+# Import from Qiskit Aer noise module
```
Unnecessary comment.

ori-pardo commented on qiskit_utils.py:
```diff
@@ -0,0 +1,173 @@
+
+def get_thermal_error_object(T1s, T2s, time_u1=0, time_u2=50, time_u3=100, time_cx=300, time_reset=1000,
+                             time_measure=1000):
+    # written by Guy, 2022_02_05
```
Delete this line

ori-pardo commented on qiskit_utils.py:
```diff
@@ -0,0 +1,173 @@
+
+
+def get_instruction_duration(gates, n_qubits, locality, duration, units='s'):
+    # constructs a list (lst) of tuples in the format that Qiskit transpile function
+    # wants:  [(instruction_name, qubits, duration, unit), …].
+    # see https://qiskit.org/documentation/stubs/qiskit.compiler.transpile.html under "instruction_durations".
+    #
+    # args:
+    # gates :  a set list of gates (strings) e.g. ['u1', 'u2', 'u3', 'cx']
+    # n_qubits: an integer
+    # locality: a list of integers of the same length as gates, storinmg the locality of each gate. the locality of a single qubit gate is 1 and the locality of a two-qubit gate is 2. for a gate with locality 2 the function will create a tuple for each possible combination of two qubits. (so doing only nearest neighbours etc. is not supported yet)
+    # duration: a list, duration for each gate.
```
Docstring in `"""`s, not with `#`. fix everywhere.

ori-pardo replied:
Use only one docstring style. fix everywhere.

ori-pardo commented on qiskit_utils.py:
```diff
@@ -0,0 +1,173 @@
+                            basis_gates=['u1', 'u2', 'u3', 'cz', 'save_density_matrix'],
+                            basis_gate_times=np.array([0, 25, 50, 26, 0]) * 1e-9,
+                            basis_gates_locality=[1, 1, 1, 2, 'all'],
+                            time_step=1e-9):
```
`1-9e` is used in lots of places. consider making this a constant

ori-pardo commented on qiskit_utils.py:
```diff
@@ -0,0 +1,173 @@
+    # basis_gates_times : duration for eahc gate
+    # basis_gates_locality:  a list of integers of the same length as gates, storinmg the locality of each gate. the locality of a single qubit gate is 1 and the locality of a two-qubit gate is 2. for a gate with locality 2 the function will create a noise model for each possible combination of two qubits. (so doing only nearest neighbours etc. is not supported yet)
+
+    from qiskit.providers.aer.noise import RelaxationNoisePass
```
imports at the head of the file. fix everywhere.
